internal/setup/setup.go

```diff
@@ -623,3 +650,134 @@ func Select(release *Release, slices []SliceKey) (*Selection, error) {
 
 	return selection, nil
 }
+
+// pathInfoToYAML converts a PathInfo object to a yamlPath object.
+func pathInfoToYAML(pi *PathInfo) (*yamlPath, error) {
+	if pi == nil {
+		return nil, nil
+	}
+	path := &yamlPath{
+		Mode:    yamlMode(pi.Mode),
+		Mutable: pi.Mutable,
+		Until:   pi.Until,
+		Arch:    yamlArch{List: pi.Arch},
+	}
+	switch pi.Kind {
+	case DirPath:
+		path.Dir = true
+	case CopyPath:
+		path.Copy = pi.Info
+	case TextPath:
+		path.Text = &pi.Info
+	case SymlinkPath:
+		path.Symlink = pi.Info
+	case GlobPath:
+	default:
+		return nil, fmt.Errorf("internal error: unrecognised PathInfo type: %s", pi.Kind)
+	}
+	return path, nil
+}
+
+// sliceToYAML converts a Slice object to a yamlSlice object.
+func sliceToYAML(s *Slice) (*yamlSlice, error) {
+	if s == nil {
+		return nil, nil
+	}
+	slice := &yamlSlice{
+		Essential: make([]string, len(s.Essential)),
+		Contents:  make(map[string]*yamlPath, len(s.Contents)),
+		Mutate:    s.Scripts.Mutate,
+	}
+	for i, key := range s.Essential {
+		slice.Essential[i] = key.String()
+	}
+	for path, info := range s.Contents {
+		pi := info
+		yamlPath, err := pathInfoToYAML(&pi)
+		if err != nil {
+			return nil, err
+		}
+		slice.Contents[path] = yamlPath
+	}
+	return slice, nil
+}
+
+// packageToYAML converts a Package object to a yamlPackage object.
+func packageToYAML(p *Package) (*yamlPackage, error) {
+	if p == nil {
+		return nil, nil
+	}
+	pkg := &yamlPackage{
+		Name:    p.Name,
+		Archive: p.Archive,
+		Slices:  make(map[string]yamlSlice, len(p.Slices)),
+	}
+	for name, slice := range p.Slices {
+		yamlSlice, err := sliceToYAML(slice)
+		if err != nil {
+			return nil, err
+		}
+		pkg.Slices[name] = *yamlSlice
+	}
+	return pkg, nil
+}
+
+// TODO These need testing
+
+var releaseExp = regexp.MustCompile(`^([a-z](?:-?[a-z0-9]){2,})-([0-9]+(?:\.?[0-9])+)$`)
+
+func parseReleaseInfo(release string) (label, version string, err error) {
+	match := releaseExp.FindStringSubmatch(release)
+	if match == nil {
+		return "", "", fmt.Errorf("invalid release reference: %q", release)
+	}
+	return match[1], match[2], nil
+}
+
+func readReleaseInfo() (label, version string, err error) {
+	data, err := os.ReadFile("/etc/lsb-release")
+	if err == nil {
+		const labelPrefix = "DISTRIB_ID="
+		const versionPrefix = "DISTRIB_RELEASE="
+		for _, line := range strings.Split(string(data), "\n") {
+			switch {
+			case strings.HasPrefix(line, labelPrefix):
+				label = strings.ToLower(line[len(labelPrefix):])
+			case strings.HasPrefix(line, versionPrefix):
+				version = line[len(versionPrefix):]
+			}
+			if label != "" && version != "" {
+				return label, version, nil
+			}
+		}
+	}
+	return "", "", fmt.Errorf("cannot infer release via /etc/lsb-release, see the --release option")
+}
+
+// GetRelease returns the release and release label (e.g. ubuntu-22.04 or
+// /path/to/release/dir/ if a directory was passed as input).
+func GetRelease(releaseStr string) (release *Release, releaseLabel string, err error) {
```

Comment:
Now that I am reading the code with the functions in the correct file I think the `GetRelease` name is a little bit confusing. We also have the `ReadRelease` function which does a very similar thing.

Sorry for not flagging this earlier.

Reply:
We could refactor this by including the label in the Release struct and then using this function as `GetRelease` and make `ReadRelease` internal (`readReleaseFromDir` or something).

Alternatively return label as well and ignore it where it is not useful. I am more inclined towards having it as a member of the struct though.

Reply:
If `Label` does not mean anything for a local directory then we should not add it. In the current implementation the label is the path, right?

We can make it `ReadOrFetchRelease`, the only thing I am worried about is having three functions that essentially do the same thing and call each other internally.

Reply:
>  In the current implementation the label is the path, right?

For `info`, yes.

Reply:
Renamed to `readOrFetchRelease` and moved back to the `main` package as per discussions offline. Changes in https://github.com/canonical/chisel/pull/101/commits/fe2e425e16282b3f4af67d25a96014cabca4fdea.